File: src/main/java/com/machfour/macros/names/EnglishColumnNames.java
package com.machfour.macros.names;

import com.machfour.macros.core.Column;
import com.machfour.macros.core.Schema.FoodTable;
import com.machfour.macros.core.Schema.NutritionDataTable;
import com.machfour.macros.objects.NutritionData;

import org.jetbrains.annotations.NotNull;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

import static com.machfour.macros.core.Schema.NutritionDataTable.*;

public class EnglishColumnNames implements ColumnNamer {
    // shortest name for each nutrient printed in default mode
    public static final Map<Column<NutritionData, Double>, String> briefNutrientNames;
    // longer name for nutrients printed in verbose mode
    public static final Map<Column<NutritionData, Double>, String> longerNutrientNames;
    // full name for each nutrient
    private static final Map<Column<NutritionData, Double>, String> prettyNutrientNames;

    private static final Map<Column<?, ?>, String> columnNames;

    static {
        Map<Column<NutritionData, Double>, String> _briefNames = new HashMap<>();
        _briefNames.put(CALORIES, "Cals");
        _briefNames.put(PROTEIN, "Prot");
        _briefNames.put(FAT, "Fat");
        _briefNames.put(CARBOHYDRATE, "Carb");
        _briefNames.put(QUANTITY, "Qty");
        briefNutrientNames = Collections.unmodifiableMap(_briefNames);
    }

    static {
        Map<Column<NutritionData, Double>, String> _longerNames = new HashMap<>();
        _longerNames.put(CALORIES, "Cals");
        _longerNames.put(PROTEIN, "Protæn");
        _longerNames.put(FAT, "Fat");
        _longerNames.put(SATURATED_FAT, "SatFat");
        _longerNames.put(CARBOHYDRATE, "Carbs");
        _longerNames.put(SUGAR, "Sugar");
        _longerNames.put(FIBRE, "Fibre");
        _longerNames.put(SODIUM, "Sodium");
        _longerNames.put(CALCIUM, "Ca");
        _longerNames.put(QUANTITY, "Qty");
        longerNutrientNames = Collections.unmodifiableMap(_longerNames);
    }

    static {
        prettyNutrientNames = new HashMap<>();
        prettyNutrientNames.put(KILOJOULES, "Kilojoules");
        prettyNutrientNames.put(CALORIES, "Calories");
        prettyNutrientNames.put(PROTEIN, "Protein");
        prettyNutrientNames.put(FAT, "Fat");
        prettyNutrientNames.put(SATURATED_FAT, "Saturated Fat");
        prettyNutrientNames.put(CARBOHYDRATE, "Carbohydrate");
        prettyNutrientNames.put(SUGAR, "Sugars");
        prettyNutrientNames.put(FIBRE, "Fibre");
        prettyNutrientNames.put(SODIUM, "Sodium");
        prettyNutrientNames.put(CALCIUM, "Calcium");
        prettyNutrientNames.put(QUANTITY, "Quantity");
        prettyNutrientNames.put(IRON, "Iron");
        prettyNutrientNames.put(STARCH, "Starch");
        prettyNutrientNames.put(CARBOHYDRATE_BY_DIFF, "Carbohydrate by difference");
        prettyNutrientNames.put(OMEGA_3_FAT, "Omega 3");
        prettyNutrientNames.put(OMEGA_6_FAT, "Omega 6");
        prettyNutrientNames.put(MONOUNSATURATED_FAT, "Monounsaturated Fat");
        prettyNutrientNames.put(POLYUNSATURATED_FAT, "Polyunsaturated Fat");
        prettyNutrientNames.put(SUGAR_ALCOHOL, "Sugar Alcohol");
        prettyNutrientNames.put(WATER, "Water");
        prettyNutrientNames.put(POTASSIUM, "Potassium");
        prettyNutrientNames.put(SALT, "Salt");
        prettyNutrientNames.put(ALCOHOL, "Alcohol");
    }

    static {
        columnNames = new HashMap<>();
        // Food Table
        columnNames.put(FoodTable.ID, "ID");
        columnNames.put(FoodTable.CREATE_TIME, "Creation time");
        columnNames.put(FoodTable.MODIFY_TIME, "Last modified");
        columnNames.put(FoodTable.INDEX_NAME, "Index name");
        columnNames.put(FoodTable.BRAND, "Brand");
        columnNames.put(FoodTable.VARIETY, "Variety");
        columnNames.put(FoodTable.VARIETY_AFTER_NAME, "Variety after name");
        columnNames.put(FoodTable.NAME, "Name");
        columnNames.put(FoodTable.NOTES, "Notes");
        columnNames.put(FoodTable.FOOD_TYPE, "Food Type");
        columnNames.put(FoodTable.USDA_INDEX, "USDA DB index");
        columnNames.put(FoodTable.NUTTAB_INDEX, "NUTTAB DB index");
        columnNames.put(FoodTable.CATEGORY, "Category");

        // NutritionData Table
        columnNames.put(NutritionDataTable.ID, "ID");
        columnNames.put(NutritionDataTable.CREATE_TIME, "Creation time");
        columnNames.put(NutritionDataTable.MODIFY_TIME, "Last modified");
        columnNames.put(NutritionDataTable.DATA_SOURCE, "Data source");
        columnNames.put(NutritionDataTable.DENSITY, "Density");
        columnNames.put(NutritionDataTable.FOOD_ID, "Food ID");
        columnNames.put(NutritionDataTable.QUANTITY_UNIT, "Quantity Unit");
        columnNames.putAll(prettyNutrientNames);
    }

    //Singleton pattern
    private EnglishColumnNames() {}
    private static EnglishColumnNames INSTANCE;
    public static EnglishColumnNames getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new EnglishColumnNames();
        }
        return INSTANCE;
    }

    @NotNull
    @Override
    public String getName(Column<?, ?> col) {
        // TODO put ALLLLLLLLLLL columns in here
        String name;
        if (columnNames.containsKey(col)) {
            name = columnNames.get(col);
        } else if (prettyNutrientNames.containsKey(col)) {
            name = prettyNutrientNames.get(col);
        } else {
            throw new UnsupportedOperationException("Name for " + col + " not yet added, sorry!");
        }
        assert name != null;
        return name;

    }

    @NotNull
    @Override
    public String getNutrientName(Column<NutritionData, Double> col) {
        if (prettyNutrientNames.containsKey(col)) {
            String name = prettyNutrientNames.get(col);
            assert name != null;
            return name;
        } else {
            throw new IllegalArgumentException("No such nutrient column: " + col);
        }
    }

}

File: src/main/kotlin/com/machfour/macros/names/EnglishColumnNames.kt
package com.machfour.macros.names

import com.machfour.macros.core.Column
import com.machfour.macros.core.Schema.FoodTable
import com.machfour.macros.core.Schema.NutritionDataTable
import com.machfour.macros.objects.NutritionData

// Singleton pattern
class EnglishColumnNames private constructor(): ColumnNamer {
    companion object {
        // shortest name for each nutrient printed in default mode
        val briefNutrientNames: Map<Column<NutritionData, Double>, String> = mapOf(
                  NutritionDataTable.CALORIES to "Cals"
                , NutritionDataTable.PROTEIN to "Prot"
                , NutritionDataTable.FAT to "Fat"
                , NutritionDataTable.CARBOHYDRATE to "Carb"
                , NutritionDataTable.QUANTITY to "Qty"
        )

        // longer (but really medium name for nutrients printed in verbose mode
        val longerNutrientNames: Map<Column<NutritionData, Double>, String> = mapOf(
                  NutritionDataTable.CALORIES to "Cals"
                , NutritionDataTable.PROTEIN to "Protæn"
                , NutritionDataTable.FAT to "Fat"
                , NutritionDataTable.SATURATED_FAT to "Sat Fat"
                , NutritionDataTable.CARBOHYDRATE to "Carbs"
                , NutritionDataTable.SUGAR to "Sugar"
                , NutritionDataTable.FIBRE to "Fibre"
                , NutritionDataTable.SODIUM to "Sodium"
                , NutritionDataTable.CALCIUM to "Ca"
                , NutritionDataTable.QUANTITY to "Qty"
        )

        // TODO put ALLLLLLLLLLL columns in here
        private val columnNames: Map<Column<*, *>, String> = mapOf(
                // Food Table
                  FoodTable.ID to "ID"
                , FoodTable.CREATE_TIME to "Creation time"
                , FoodTable.MODIFY_TIME to "Last modified"
                , FoodTable.INDEX_NAME to "Index name"
                , FoodTable.BRAND to "Brand"
                , FoodTable.VARIETY to "Variety"
                , FoodTable.EXTRA_DESC to "Extra description"
                , FoodTable.NAME to "Name"
                , FoodTable.NOTES to "Notes"
                , FoodTable.FOOD_TYPE to "Food Type"
                , FoodTable.USDA_INDEX to "USDA DB index"
                , FoodTable.NUTTAB_INDEX to "NUTTAB DB index"
                , FoodTable.CATEGORY to "Category"

                // NutritionData Table
                , NutritionDataTable.ID to "ID"
                , NutritionDataTable.CREATE_TIME to "Creation time"
                , NutritionDataTable.MODIFY_TIME to "Last modified"
                , NutritionDataTable.DATA_SOURCE to "Data source"
                , NutritionDataTable.DENSITY to "Density"
                , NutritionDataTable.FOOD_ID to "Food ID"
                , NutritionDataTable.QUANTITY_UNIT to "Quantity Unit"

                // Nutrient columns
                , NutritionDataTable.KILOJOULES to "Kilojoules"
                , NutritionDataTable.CALORIES to "Calories"
                , NutritionDataTable.PROTEIN to "Protein"
                , NutritionDataTable.FAT to "Fat"
                , NutritionDataTable.SATURATED_FAT to "Saturated Fat"
                , NutritionDataTable.CARBOHYDRATE to "Carbohydrate"
                , NutritionDataTable.SUGAR to "Sugars"
                , NutritionDataTable.FIBRE to "Fibre"
                , NutritionDataTable.SODIUM to "Sodium"
                , NutritionDataTable.CALCIUM to "Calcium"
                , NutritionDataTable.QUANTITY to "Quantity"
                , NutritionDataTable.IRON to "Iron"
                , NutritionDataTable.STARCH to "Starch"
                , NutritionDataTable.CARBOHYDRATE_BY_DIFF to "Carbohydrate by difference"
                , NutritionDataTable.OMEGA_3_FAT to "Omega 3"
                , NutritionDataTable.OMEGA_6_FAT to "Omega 6"
                , NutritionDataTable.MONOUNSATURATED_FAT to "Monounsaturated Fat"
                , NutritionDataTable.POLYUNSATURATED_FAT to "Polyunsaturated Fat"
                , NutritionDataTable.SUGAR_ALCOHOL to "Sugar Alcohol"
                , NutritionDataTable.WATER to "Water"
                , NutritionDataTable.POTASSIUM to "Potassium"
                , NutritionDataTable.SALT to "Salt"
                , NutritionDataTable.ALCOHOL to "Alcohol"
        )

        val instance: EnglishColumnNames = EnglishColumnNames()
    }

    @Suppress("UNCHECKED_CAST")
    override fun getName(col: Column<*, *>): String {
        return columnNames[col]
                ?: throw UnsupportedOperationException("Name for $col not yet added, sorry!")
    }

}